lib/seattleflu/id3c/cli/command/etl/redcap_det_uw_reopening.py

```diff
@@ -0,0 +1,1312 @@
+
+    if not barcode:
+        LOG.warning("Could not create Specimen Resource due to lack of barcode.")
+        return None, None
+
+    specimen_identifier = create_identifier(
+        system = f"{INTERNAL_SYSTEM}/sample",
+        value = barcode
+    )
+
+    # YYYY-MM-DD HH:MM:SS in REDCap
+    received_time = record['samp_process_date'].split()[0] if record.get('samp_process_date') else None
+
+    note = None
+
+    if record['able_to_test'] == 'no':
+        note = 'never-tested'
+    else:
+        note = 'can-test'
+
+    specimen_type = 'NSECR'  # Nasal swab.
+    specimen_resource = create_specimen_resource(
+        [specimen_identifier], patient_reference, specimen_type, received_time,
+        collection_date(record, collection_method), note
+    )
+
+    return create_entry_and_reference(specimen_resource, "Specimen")
+
+
+def collection_date(record: dict, collection_method: CollectionMethod) -> Optional[str]:
+    """
+    Determine sample/specimen collection date from the given REDCap *record*.
+    """
+    if collection_method == CollectionMethod.KIOSK:
+        return record["nasal_swab_q"]
+    elif collection_method == CollectionMethod.SWAB_AND_SEND:
+        return record["date_on_tube"] or record["kit_reg_date"] or record["back_end_scan_date"]
+    else:
+        return None
+
+def combine_multiple_fields(record: Dict[Any, Any], field_prefix: str, field_suffix: str = "") -> Optional[List]:
+        """
+        Handles the combining of multiple fields asking the same question such
+        as country and state traveled.
+        """
+        regex = rf'^{re.escape(field_prefix)}[0-9]+{re.escape(field_suffix)}$'
+        empty_value = ''
+        answered_fields = list(filter(lambda f: filter_fields(f, record[f], regex, empty_value), record))
+
+        if not answered_fields:
+            return None
+
+        return list(map(lambda x: record[x], answered_fields))
+
+def create_enrollment_questionnaire_response(record: dict, patient_reference: dict,
+                                            encounter_reference: dict) -> Optional[dict]:
+    """
+    Returns a FHIR Questionnaire Response resource entry for the enrollment
+    encounter (i.e. encounter of enrollment into the study)
+    """
+
+    # Do not include `core_age_years` because we calculate the age ourselves in the computed questionnaire.
+    integer_questions = [
+        'core_zipcode',
+        'weight',
+        'height_ft',
+        'height_in',
+        'height_ft_calc',
+        'height_total',
+        'vaccine_month',
+        'vaccine_year'
+    ]
+
+    string_questions = [
+        'study_area',
+        'text_or_email',
+        'text_or_email_attestation',
+        'campus_location',
+        'student_or_employee',
+        'student_level',
+        'sea_employee_type',
+        'uw_greek_house',
+        'core_home_state',
+        'core_house_members',
+        'dorm',
+        'uw_apt_names',
+        'core_sex',
```

Comment:
Gender is already part of the the patient resource so no need to ingest this field again.

Reply:
The theme will my responses to all these is that just it's someplace else related to the encounter, but for convenience I want to ingest it here too.

Reply:
What do you mean by convenience?

Reply:
Convenience when someone wants to look at the questionnaire responses. The symptoms question is a question in the questionnaire. Yes, it happens to be symptoms and we process those elsewhere, but the symptoms question is a question in the questionnaire.

Reply:
All data we collect comes from the questionnaire, but we add them to the appropriate FHIR resources. We use the QuestionnaireResponse resource to hold data that does not fit into any other FHIR resource.

Reply:
You seem passionate that I should not include the symptom questions as questionnaire items because we store them as symptoms. My point is that they are questions in the surveys, so why not include them as questionnaire items? I'll take them out if this if this really bothers you.

Reply:
> We use the QuestionnaireResponse resource to hold data that does not fit into any other FHIR resource.

Jover's right. We want our FHIR bundles to be as clean as possible. The definition of [data cleansing](https://en.wikipedia.org/wiki/Data_cleansing) includes removing duplicate representations of the same entity. Therefore, we should only represent gender, symptoms, etc. once in a Bundle for each encounter.

Reply:
I'm no longer ingesting the symptom questions or core_sex as questionnaire items.